mesh/meshdb.go

```diff
@@ -327,30 +350,159 @@ func (m *MeshDB) writeTransactions(txs []*types.AddressableSignedTransaction) er
 	return nil
 }
 
-func (m *MeshDB) GetTransactions(transactions []types.TransactionId) (
-	map[types.TransactionId]*types.AddressableSignedTransaction,
-	[]types.TransactionId) {
+func (m *MeshDB) addToMeshTxs(txs []*types.Transaction, layer types.LayerID) error {
+	groupedTxs := groupByOrigin(txs)
 
-	var mIds []types.TransactionId
-	ts := make(map[types.TransactionId]*types.AddressableSignedTransaction, len(transactions))
-	for _, id := range transactions {
-		t, err := m.GetTransaction(id)
-		if err != nil {
-			m.Warning("could not fetch tx, %v %v", hex.EncodeToString(id[:]), err)
-			mIds = append(mIds, id)
-		} else {
-			ts[id] = t
+	for addr, accountTxs := range groupedTxs {
+		if err := m.addToAccountTxs(addr, accountTxs, layer); err != nil {
+			return err
 		}
 	}
-	return ts, mIds
+	return nil
+}
+
+func (m *MeshDB) addToAccountTxs(addr types.Address, accountTxs []*types.Transaction, layer types.LayerID) error {
+	m.meshTxsMutex.Lock()
+	defer m.meshTxsMutex.Unlock()
+
+	// TODO: instead of storing a list, use LevelDB's prefixed keys and then iterate all relevant keys
+	pending, err := m.getAccountPendingTxs(addr)
+	if err != nil {
+		return err
+	}
+	pending.Add(layer, accountTxs...)
+	if err := m.storeAccountPendingTxs(addr, pending); err != nil {
```

Comment:
why account pending txs need to be persistent?

Reply:
The mesh is persistent and they have to match what's already in the mesh, since we only update it when updating the mesh.